ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/commons/objectivefunctionevaluator/RemedialActionCostEvaluator.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2024, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.powsybl.openrao.searchtreerao.commons.objectivefunctionevaluator;
+
+import com.powsybl.openrao.commons.Unit;
+import com.powsybl.openrao.data.cracapi.State;
+import com.powsybl.openrao.data.cracapi.cnec.FlowCnec;
+import com.powsybl.openrao.data.cracapi.rangeaction.HvdcRangeAction;
+import com.powsybl.openrao.data.cracapi.rangeaction.InjectionRangeAction;
+import com.powsybl.openrao.data.cracapi.rangeaction.PstRangeAction;
+import com.powsybl.openrao.data.cracapi.rangeaction.RangeAction;
+import com.powsybl.openrao.raoapi.parameters.RangeActionsOptimizationParameters;
+import com.powsybl.openrao.searchtreerao.result.api.FlowResult;
+import com.powsybl.openrao.searchtreerao.result.api.RemedialActionActivationResult;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * @author Thomas Bouquet {@literal <thomas.bouquet at rte-france.com>}
+ */
+public class RemedialActionCostEvaluator implements CostEvaluator {
+    private final Set<State> optimizedStates;
+    private final Set<FlowCnec> flowCnecs;
+    private final Unit unit;
+    private final MarginEvaluator marginEvaluator;
+    private final RangeActionsOptimizationParameters rangeActionsOptimizationParameters;
+
+    public RemedialActionCostEvaluator(Set<State> optimizedStates, Set<FlowCnec> flowCnecs, Unit unit, MarginEvaluator marginEvaluator, RangeActionsOptimizationParameters rangeActionsOptimizationParameters) {
+        this.optimizedStates = optimizedStates;
+        this.flowCnecs = flowCnecs;
+        this.unit = unit;
+        this.marginEvaluator = marginEvaluator;
+        this.rangeActionsOptimizationParameters = rangeActionsOptimizationParameters;
+    }
+
+    @Override
+    public String getName() {
+        return "remedial-action-cost-evaluator";
+    }
+
+    @Override
+    public Pair<Double, List<FlowCnec>> computeCostAndLimitingElements(FlowResult flowResult, RemedialActionActivationResult remedialActionActivationResult, Set<String> contingenciesToExclude) {
```

Comment:
I agree it's starting to get a bit confusing. I think we did this to avoid recomputing an ordered list on all cnecs every time we want to have to get the most limiting element(s) for any reason (like logging). But we could still store that list in each relevant evaluator object, and add a getMostConstrainedElements method or something like that to return the wanted info.

Reply:
I agree that we should at least split the method in two for a better clarity